### PR TITLE
Added GFX shapes to renderer

### PR DESCRIPTION
**Pull request by Peter-JanGootzen**
This PR was created by Bram-Boris and me during the work on our [DPA/AGLA assessment](https://github.com/Peter-JanGootzen/FlatGalaxySociety). To use the engine we had needed circle and line shapes, so we added those to the renderer. Sadly we had to slight rework the renderer to work with polymorph renderables and not renderables that are not textures. Polymorph renderables is something we were probably going to need at a later stage anyway with Beast Arena.

### How does it work?
We basicly applied the visitor pattern here. We have the renderer who calls the render function on every renderable and those renderables call the render function corresponding to their own type on .the renderer.

### How do I test this?
Use this version of the gameloop
```c++
void GameController::gameLoop() {
    BrickEngine engine = BrickEngine("Beast Arena", 1280, 720, { 1, 2, 3 });
    this->renderable_factory = engine.getRenderableFactory();
    // Always draw the FPS counter on the highest layer
    this->top_layer = 2;
    this->createFpsCounter(0);

    auto dst_max = std::unique_ptr<Rect>(new Rect { 0, 0, 1000, 330});
    std::unique_ptr<Renderable> makker_max = renderable_factory->createImage("./assets/graphics/verstappen_laughing.bmp", 1, std::move(dst_max));

    auto text = this->renderable_factory->createText("TESTEST", 200, { 255, 255, 255, 0 }, this->top_layer, std::unique_ptr<Rect>(new Rect {0, 0, 400, 400}));

    auto circle = renderable_factory->createCircle(600, 350, 300, true, { 255, 255, 1, 200 }, 2);
    auto line = renderable_factory->createLine(50, 600, 50, 50, { 255, 20, 255, 200 }, 2);

    Renderer* renderer = engine.getRenderer();

    int i = 0;
    while(true) {
        std::cout << i << std::endl;
        int start_time = engine.getTicks();

        renderer->clearScreen();
        renderer->queueRenderable(line.get());
        renderer->queueRenderable(circle.get());
        renderer->queueRenderable(fps_counter.get());
        renderer->queueRenderable(text.get());
        renderer->queueRenderable(makker_max.get());
        renderer->drawScreen();

        int fps = this->calculateFps(&engine, start_time);
        this->createFpsCounter(fps);
        i++;
    }
}
```

### Other things
- We fixed the rendering bug when putting multiple things on the same layer where only the last queued renderable would get rendered, this was because Luuk and Peter-Jan didn't know unordererd_map could only have a single instance of a key.
- Some uses of the unique_ptr constructor were changed to make_unique, because this is supposed to be better. [source](https://stackoverflow.com/questions/37514509/advantages-of-using-stdmake-unique-over-new-operator)

### Bugs introduced
- After drawing GFX shapes, the very last _text_ renderable that gets drawn will have a rogue pixel in the bottom right of it's render destination size. We were not able to figure out how to fix this. Underneath is an image that shows the bug, you should probably zoom in to see the single pixel. ;)
2019-10-04-220007_1280x720_scrot

### CODE REVIEW

File: include/brickengine/rendering/renderables/circle.hpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
+#ifndef FILE_CIRCLE_HPP
+#define FILE_CIRCLE_HPP
+
+#include "brickengine/rendering/renderables/renderable.hpp"
+#include "brickengine/rendering/renderables/circle.hpp"
+#include "brickengine/rendering/renderer.hpp"
+
+class Renderer;
```
[the remark this answers is not in the document]
Comment (Peter-JanGootzen) [reply include/brickengine/rendering/renderables/circle.hpp]: This is called a forward declaration to fix cyclic depedency. The second CPP1 lecture explained this. https://stackoverflow.com/questions/625799/resolve-build-errors-due-to-circular-dependency-amongst-classes

File: include/brickengine/rendering/renderables/circle.hpp [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+class Circle : public Renderable {
+public:
+    Circle(int x, int y, int radius, bool filled, Color color, int layer);
```
[the remark this answers is not in the document]
Comment (Peter-JanGootzen) [reply include/brickengine/rendering/renderables/circle.hpp]: These values are copied over to the Circle, making them const won't do anything.

File: include/brickengine/rendering/renderables/texture.hpp [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+class Renderer;
+
+class Texture : public Renderable {
+public:
+    Texture(SDL_Texture* texture, int layer, std::unique_ptr<Rect> dst);
+    Texture(SDL_Texture* texture, int layer, std::unique_ptr<Rect> dst, std::unique_ptr<Rect> src);
+    void render(Renderer& r);
+    virtual Rect* getSrcRect();
+    virtual Rect* getDstRect();
+    virtual SDL_Texture* getTexture();
```
[the remark this answers is not in the document]
Comment (Peter-JanGootzen) [reply include/brickengine/rendering/renderables/texture.hpp]: True dat

File: include/brickengine/rendering/renderables/line.hpp [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+#include "brickengine/rendering/renderables/renderable.hpp"
+#include "brickengine/rendering/renderables/data/color.hpp"
+#include "brickengine/rendering/renderer.hpp"
+
+class Renderer;
+
+class Line : public Renderable {
+public:
+    Line(int x1, int y1, int x2, int y2, Color color, int layer);
```
[the remark this answers is not in the document]
Comment (Peter-JanGootzen) [reply include/brickengine/rendering/renderables/line.hpp]: Same case as with the Circle. These values are copied over to the line, making them const won't do anything.

File: include/brickengine/rendering/renderables/circle.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (markJavab) [reply include/brickengine/rendering/renderables/circle.hpp]: Thank you, totally forgot about this

File: lib/rendering/renderables/texture.cpp [hunk 5]
```diff
@@ -0,0 +1,34 @@
+#include <memory>
+#include "SDL2/SDL.h"
+#include "brickengine/rendering/renderables/texture.hpp"
+#include "brickengine/rendering/renderer.hpp"
+
+void destroyT(SDL_Texture* t) {
+    SDL_DestroyTexture(t);
+}
```
Comment (LuukSantegoeds) [inline lib/rendering/renderables/texture.cpp]: Is het niet duidelijker als je dit _destroyTexture_ noemt?

**Pull request merged**